src/ray/pubsub/publisher.cc

```diff
@@ -20,149 +20,188 @@ namespace pubsub {
 
 namespace pub_internal {
 
-void SubscriptionIndex::AddEntry(const ObjectID &object_id,
-                                 const SubscriberID &subscriber_id) {
-  auto &subscribing_objects = subscribers_to_objects_[subscriber_id];
-  RAY_CHECK(subscribing_objects.emplace(object_id).second);
-  auto &subscriber_map = objects_to_subscribers_[object_id];
+template <typename MessageID>
+void SubscriptionIndex<MessageID>::AddEntry(const std::string &message_id_binary,
+                                            const SubscriberID &subscriber_id) {
+  const auto message_id = ParseBinary(message_id_binary);
+  auto &subscribing_message_ids = subscribers_to_message_id_[subscriber_id];
+  RAY_CHECK(subscribing_message_ids.emplace(message_id).second);
+  auto &subscriber_map = message_id_to_subscribers_[message_id];
   RAY_CHECK(subscriber_map.emplace(subscriber_id).second);
 }
 
+template <typename MessageID>
 absl::optional<std::reference_wrapper<const absl::flat_hash_set<SubscriberID>>>
```

Comment:
I'd prefer returning raw pointer instead. std::ref makes things look complicated. 
```
const absl::flat_hash_set<SubscriberID>*
```
return nullptr if not exist. For optional kind of thing, you always need to check whether it exists or not. Usually, if it's passing-by-value, and it can also be null, I'll use optional.